test/Microsoft.ML.TestFramework/RemoteExecutorTestBase.cs

```diff
@@ -0,0 +1,243 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Sdk;
+
+namespace Microsoft.ML.TestFramework
+{
+    /// <summary>Base class used for all tests that need to spawn a remote process.</summary>
+    public abstract partial class RemoteExecutorTestBase
```

Comment:
> we can easily run them on ci

The same way that CoreCLR is running them. You have a separate leg for each configuration you want to test.

> There is a separate test for each code path and a person easily find and debug if any path is misbehaving.

Yes, this is better at that regards. The downside is that each test now has to have a `RemoteExecuteTestBase` call and wrap the logic in a delegate.

We could instead take advantage of various extensibility features that xUnit has. You would just need to define a custom fact/theory-attribute/discoverer. A simple example of this is the `RetryFact`: https://github.com/xunit/samples.xunit/tree/master/RetryFactExample

You would then just configure the `RunAsync` method in the custom `*TestCase` to kick off a process with the appropriate environment variables (basically what `RemoteExecutorTestBase` is doing).

This would ultimately mean that the test code continues looking "normal" and all the environment set-up/remote execution would be abstracted away.

Reply:
i was able to add a retryFact conditionalTheory attribute.
The problem i am facing in how to Kickoff RunAsync method in a different process ?

here is the sample code https://github.com/Anipik/machinelearning/commit/be09268d30e10acf8b3a6168220333f6a848f5fe

Reply:
What Tanner suggests would be ideal - I didn't realize that there was an extensibility point that would allow something like this - and I assume @stephentoub didn't know about it either when he wrote the RemoteExec infrastructure originally. However, with async and cross process in the picture it could be a bit fiddly to get this right. I would rather not do that in this PR. I suggest to use the tried and tested RemoteExec pattern we use in CoreFX here. If we want to improve it, we might want to do that in CoreFX first and then copy it here.

Reply:
@eerhardt thoughts? Even if we do this it should be in a different PR, in my opinion.

Reply:
Doing this in a separate PR sounds fine with me.